### PR TITLE
KeyPressManager now manages listener (and cityscreen usage)

### PR DESCRIPTION
**Pull request by SomeTroglodyte**
Replacing #3933 - this diff looks how I remember it...

### CODE REVIEW

File: core/src/com/unciv/ui/utils/KeyPressDispatcher.kt [hunk 1]
```diff
@@ -61,14 +82,97 @@ class KeyPressDispatcher: HashMap<KeyCharAndCode, (() -> Unit)>() {
...
+        installStage = null
+    }
+
+    /** Implements lazy hooking of the listener into the stage.
+     *
+     *  The listener will be added to the stage's listeners only when and as soon as [this] contains mappings.
+     *  When all mappings are removed or cleared the listener is removed from the stage.
+     */
+    private fun checkInstall(forceRemove: Boolean = false) {
+        if (listener != null && installStage != null) {
```
Comment (yairm210) [inline core/src/com/unciv/ui/utils/KeyPressDispatcher.kt:158]: Please reverse if to exit quickly and remove indentation :)

File: core/src/com/unciv/ui/utils/KeyPressDispatcher.kt [hunk 2]
```diff
@@ -61,14 +82,97 @@ class KeyPressDispatcher: HashMap<KeyCharAndCode, (() -> Unit)>() {
...
     fun revertToCheckPoint() {
         keys.minus(checkpoint).forEach { remove(it) }
+        checkInstall()
+    }
+
+    /** install our [EventListener] on a stage with optional inhibitor
+     * @param   stage The [Stage] to add the listener to
+     * @param   hasOpenPopups An optional lambda - when it returns true all keys are ignored
+     */
+    fun install(stage: Stage, name: String? = null, hasOpenPopups: (() -> Boolean)? = null) {
```
Comment (yairm210) [inline core/src/com/unciv/ui/utils/KeyPressDispatcher.kt:123]: please change lambda to ignoreAllKeys or something

File: core/src/com/unciv/ui/utils/KeyPressDispatcher.kt [hunk 3]
```diff
@@ -61,14 +82,97 @@ class KeyPressDispatcher: HashMap<KeyCharAndCode, (() -> Unit)>() {
...
+        this.name = name
+        if (installStage != null) uninstall()
+        listener =
+            object : InputListener() {
+                override fun keyTyped(event: InputEvent?, character: Char): Boolean {
+                    val key = KeyCharAndCode(event, character)
+
+                    if (!contains(key) || (hasOpenPopups?.invoke() == true))
+                        return super.keyTyped(event, character)
```
Comment (yairm210) [inline core/src/com/unciv/ui/utils/KeyPressDispatcher.kt:132]: This is just 'return false', but this makes it look like the super actually has extra logic - please change to reduce apparent complexity

File: core/src/com/unciv/ui/utils/KeyPressDispatcher.kt (hunk 1 above)
Comment (SomeTroglodyte) [reply core/src/com/unciv/ui/utils/KeyPressDispatcher.kt:158]: Another personal preference - return before the end of a function still 'tastes' a bit like a goto - and G0t0 was not the best NPC in KOTOR2... 
Uh - I notice I left some `println`'s active - shouldn't have been. They have been really useful, but should be under a compiler conditional - and I haven't learned those for this ecosystem. `#ifdef` ? :wink:

File: core/src/com/unciv/ui/utils/KeyPressDispatcher.kt (hunk 2 above)
Comment (SomeTroglodyte) [reply core/src/com/unciv/ui/utils/KeyPressDispatcher.kt:123]: Ah - of course!

File: core/src/com/unciv/ui/utils/KeyPressDispatcher.kt (hunk 3 above)
Comment (SomeTroglodyte) [reply core/src/com/unciv/ui/utils/KeyPressDispatcher.kt:132]: Uh, that's not how I see it - I mean the line is even auto-inserted by the IDE like that. Point is, super is part of an API and out of our control and subject to change - potentially. remotely possible. perhaps. So we treat it like documented - if we feel responsible we handle the event if not we let it bubble. When I think of dotnet UI frameworks like Windows Forms or WPF it was always quite vital to handle only what's needed and pass on the rest... Am I glad I've avoided those for a while now.

... or am I missing something?

**Comment by SomeTroglodyte** [comment]
'Problems' tab wasn't empty, modified a comment - let's see whether we agree this is clean enough to serve for a while.

File: core/src/com/unciv/ui/utils/KeyPressDispatcher.kt (hunk 3 above)
Comment (yairm210) [reply core/src/com/unciv/ui/utils/KeyPressDispatcher.kt:133]: I understand what you're saying. I don't necessarily agree, but I'm willing to accept this is a difference of opinion that's not critical :)

**Comment by yairm210** [comment]
Okay! I think this is ready to go, let's give it a run :)

**Comment by SomeTroglodyte** [comment]
Thanks - which means more work as the follow-up was ready and now needs rebasing... What was it? Keys in Popups... Not today.

Say, was there a specific reason there's N yes/no questions with Yes and No horizontally arranged but two which do it vertically?

**Pull request merged**